slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/ValidatorsDao.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.web3signer.slashingprotection;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.tuweni.bytes.Bytes;
+import org.jdbi.v3.sqlobject.config.RegisterBeanMapper;
+import org.jdbi.v3.sqlobject.customizer.BindList;
+import org.jdbi.v3.sqlobject.statement.SqlBatch;
+import org.jdbi.v3.sqlobject.statement.SqlQuery;
+import org.jdbi.v3.sqlobject.transaction.Transaction;
+
+public interface ValidatorsDao {
+
+  @Transaction
+  default void registerMissingValidators(final List<Bytes> validators) {
+    final List<Validator> registeredValidators = retrieveValidators(validators);
+    final List<Bytes> validatorsMissingFromDb = new ArrayList<>(validators);
+    registeredValidators.forEach(v -> validatorsMissingFromDb.remove(v.getPublicKey()));
+    registerValidators(validatorsMissingFromDb);
```

Comment:
if we are using latest version of PostgreSQL (BUT it won't be compatible with other database), we can potentially use `INSERT INTO .... ON CONFLICT DO NOTHING`. which would avoid two calls from database.

Reply:
we can't use that with an auto-incrementing id though. but perhaps we can increment the id ourselves as long as we have the appropriate transaction isolation is enabled.